Add sbt Migration tutorial (#91)
scalacenter/scala-3-migration-guide: docs/tutorials/sbt-migration.md

#### [Scala.js](https://www.scala-js.org/)

The Scala.js compilation on Scala 3 does not rely on a compiler plugin anymore.

To compile your Scala.js project in Scala 3 you need the `sbt-scalajs` plugin version `1.3.0` or higher.

```scala
// project/plugins.sbt
addSbtPlugin("org.scala-js" % "sbt-scalajs" % "1.3.0")
```

#### [Scala Native](https://scala-native.readthedocs.io/en/v0.3.9-docs/)

Scala 3 does not yet support Scala Native.

If you project is cross-built to Scala Native, you can still migrate it to Scala 3 but you have to deactivate the Scala Native compilation on Scala 3.
How to? [Contribution welcome!](../contributing.md)

#### [Kind Projector](https://github.com/typelevel/kind-projector)

The `kind-projector` plugin is not available in Scala 3.

However, Scala 3 introduces some features that makes `kind-projector` not needed in many cases:
- [Type Lambdas](http://dotty.epfl.ch/docs/reference/new-types/type-lambdas.html)
- [Kind Polymorphism](http://dotty.epfl.ch/docs/reference/other-new-features/kind-polymorphism.html)
- Poly Function

Additionally, the `-Ykind-projector` compiler option allows the usage of `*` as a wildcard in type projections.

> ### Runtime reflection

Reviewer: ```suggestion
### Run-time reflection
```

`scala-reflect` will not be ported to Scala 3 because:

> - It exposes Scala 2 compiler internals that does not exist in Scala 3.

Reviewer: ```suggestion
- It exposes Scala 2 compiler internals that do not exist in Scala 3.
```

> - Runtime reflection is less efficient than compile time reflection.

Reviewer: ```suggestion
- Run-time reflection is less efficient than compile-time reflection.
```

Reviewer: That is actually a reason for which scala-reflect will not be reimplemented. The first bullet is enough.

If your project depends on `scala-reflect`, or consumes instances of the `Manifest` class, it cannot be compiled by the Scala 3 compiler.
To remedy this situation, you can try to re-implement the corresponding parts of the code, using Java reflection or the [Scala 3 metaprogramming features](../macros/metaprogramming.md).

If `scala-reflect` is transitively added in the class path by a library, you probably need to upgrade that dependency to a Scala 3 compatible version of it.

## 2. Configure sbt for Scala 3

Support for Scala 3 in sbt is brought by the `sbt-dotty` plugin.

> This plugin provides sbt with the ability to invoke the Scala 3 compiler incrementally.

Reviewer: ```suggestion
This plugin provides sbt with the ability to invoke the Scala 3 compiler.
```
I understand what you mean with the "incrementally", but it's probably too confusing for a casual reader who's not fairly knowledgeable about how sbt invokes the compiler.

## 3. Prepare the dependencies

The fact that a Scala 3.0 module can depend on a Scala 2.13 artifact, and vice versa, is unprecedented in the history of the Scala language.
As we will see, it gives a lot more flexibility in terms of dependency management.

Using `%%` for dealing with Scala dependencies has been the norm for quite some time. A dependency on a Scala library is traditionally written:

```scala
// build.sbt
libraryDependencies += "org.typelevel" %% "cats-core" % "2.1.1"
```

The `%%` operator means that the library is dependent on the Scala version:
- if the Scala version is 2.12, the `cats-core_2.12-2.1.1.jar` artifact will be resolved
- if the Scala version is 2.13, the `cats-core_2.13-2.1.1.jar` artifact will be resolved

But suppose I switch the `scalaVersion` from `2.13.3` to `3.0.0-M1`.
When I try to compile, sbt complains that `cats-core_3.0-2.1.1.jar` cannot be found.
Of course it is right, yet I must be able to use `cats-core_2.13-2.1.1.jar` to compile to Scala 3.

### A Scala 3.0 module depending on a Scala 2.13 artifact

They are two ways to configure a Scala 3 module that depends on a Scala 2.13 artifact:

> - We can precise the binary version manually:

Reviewer: ```suggestion
- We can specify the binary version manually:
```

```scala
libraryDependencies += "org.typelevel" % "cats-core_2.13" % "2.1.1"
```

> Remark that the `%%` operator is replaced by the `%` operator.

Reviewer: ```suggestion
Note that the `%%` operator is replaced by the `%` operator.
```

> The reason is we do not need the automatic resolution of the binary version.

Reviewer: That is a bit confusing. The reason is stronger: we must not use it (not just we don't need it). Perhaps rephrase as:
```
Manually using `_2.13` replaces the behavior of `%%`.
```
or
```
Using `%%` in addition to the explicit `_2.13` would end up looking for `cast-core_2.13_2.13`, which does not exist.

Author: I opted for
```
Otherwise sbt would try to resolve `cats-core_2.13_2.13-2.1.1` which is nonsense.
```

- We can use the `withDottyCompat` extension method brought by `sbt-dotty`:
```scala
libraryDependencies +=
  ("org.typelevel" %% "cats-core" % "2.1.1").withDottyCompat(scalaVersion.value)
```
This method changes the behavior of the `%%` operator on Scala 3 versions only, as described in the below table:

| scalaVersion | binary version |
|-|-|
| `2.12.x` | `_2.12` |
| `2.13.x` | `_2.13` |
| `3.0.x` | `_2.13` |

This is much useful in a repository that is cross-built on many versions.

> ### A Scala 2.13 module depending on a Scala 3.0 artifact

Reviewer 2: For now, we also need to add the `-Ytasty-reader` flag

Author: Was it decided recently? Where can I find info?

Reviewer 2: yeah have a look at https://github.com/scala/scala/pull/9293

Conversely a Scala 2.13 module can depend on a Scala 3.0 artifact.

We can configure it so by specifying the binary version manually:

```scala
libraryDependencies += "org.typelevel" % "cats-core_2.13" % "2.1.1"
```

> Again you can remark that the `%%` is not used.

Reviewer: ```suggestion
Again you can notice that the `%%` is not used.
```
Technically, "remark" exists as a verb in English, but it's almost never used that way in modern speech.

(not a native speaker; I could be contradicted by a native speaker)

### Macro dependencies

Macro dependencies are still tied to a particular version of the compiler.
In such case the `%%` operator is still very much relevant.

For instance, the `sourcecode` library is a cross-published macro library and thus it can be configured this way:
```scala
libraryDependency += "com.lihaoyi" %% "sourcecode" % "0.2.1"
```

### What about Scala.js dependencies?

The `%%%` operator is used in Scala.js projects to automatically resolve artifacts, depending on the Scala and Scala.js versions.

It can be combined with the `withDottyCompat` extension method, as in:
```scala
libraryDependencies +=
  ("org.typelevel" %%% "cats-core" % "2.1.1").withDottyCompat(scalaVersion.value)
```

As an alternative it is possible to specify the Scala.js binary version manually:
```scala
libraryDependencies +=

>   "org.typelevel" % "cats-core_sjs1_2.13" % "2.1.1"

Reviewer: That works for a JS-only project, but not for a cross-project. In a cross-project you would need the very ugly
```scala
s"cast-core${if (crossProjectPlatfor.value == JSPlatform) "_sjs1" else ""}_2.13"
```

```

### Summing up

An sbt project can use different dependency resolution strategies, based on the binary version of the targeted libraries.

```scala
// build.sbt
val foo = project
  .settings(
    crossScalaVersions ++= Seq("2.13.3", "3.0.0-M1")
    libraryDependencies ++= Seq(
      "org.typelevel" % "cats-core_2.13" % "2.1.1",
      "com.lihaoyi" % "os-lib_3.0" % "0.7.1",
      "com.lihaoyi" %% "sourcecode" % "0.2.1"
    )
  )
```

This sbt configuration is valid on both Scala `2.13.3` and Scala `3.0.0-M1`. 
_(Not quite true, because `os-lib` and `sourcecode` are not yet published on `3.0.0-M1`)_

> The below table shows which artifacts are resolved depending on the Scala version:

Reviewer: ```suggestion
The table below shows which artifacts are resolved depending on the Scala version:
```

| scalaVersion | cats-core | os-lib | sourcecode |
|-|-|-|-|
| `2.13.3` | `_2.13` | `_3.0` | `_2.13` |
| `3.0.0-M1` | `_2.13` | `_3.0` | `_3.0` |

## 4. Set up cross-building

The two main challenges of the codebase migration are:
- Make the code compile

> - Make sure that the runtime behavior is unchanged

Reviewer: ```suggestion
- Make sure that the run-time behavior is unchanged
```

For doing so, we recommend the cross-building strategy, that is to make the code compile in Scala 3.0 while maintaining the compatibility with Scala 2.13.
While this can be tricky sometime, it gives us the ability to fallback to 2.13 and run the tests at any stage of the migration.

> That way we can fix the incompatibilities one at a time and be confident that the runtime behavior is unchanged.

Reviewer: ```suggestion
That way we can fix the incompatibilities one at a time and be confident that the run-time behavior is unchanged.
```
"runtime" is a noun that designates a piece of software such as a VM or a core library. "run-time" is an adjective that refers to "at the time of running".

The rule of thumb to know which one to use is: can you replace "run(-)time" by "compile-time" without breaking the grammatical structure of the sentence? If yes, it's "run-time"; if not, it's "runtime".

Author: Thanks for the explanation! I was really confused about it.
